File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/KaleidoDiagnostic.java
package com.tongsr.kaleido.gradle;

import org.gradle.api.GradleException;

record KaleidoDiagnostic(
        String code,
        String project,
        String variant,
        String stage,
        String origin,
        String target,
        String reason,
        String repair) {
    GradleException failure() {
        return new GradleException(render());
    }

    String render() {
        return "%s project=%s variant=%s stage=%s origin=%s target=%s reason=%s repair=%s"
                .formatted(code, project, variant, stage, origin, target, reason, repair);
    }
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/KaleidoDiagnostic.kt
package com.tongsr.kaleido.gradle

import org.gradle.api.GradleException

internal data class KaleidoDiagnostic(
    val code: String,
    val project: String,
    val variant: String,
    val stage: String,
    val origin: String,
    val target: String,
    val reason: String,
    val repair: String,
) {
    fun failure(): GradleException = GradleException(render())

    fun render(): String =
        "%s project=%s variant=%s stage=%s origin=%s target=%s reason=%s repair=%s"
            .format(code, project, variant, stage, origin, target, reason, repair)
}
